### ISSUE TITLE
Q: S3 get object parallel call seems not really parallel?

### ISSUE THREAD

**Issue by ottokruse** [aio-libs/aiobotocore#651]
Hope you can enlighten me. I wrote a little Lambda to test out aiobotocore, but I don't understand the behavior I see.

The Lambda fetches 5 files from s3, and runs in around 150ms consistently (besides the cold start run).
When I make it 10 files, I see it takes around 300ms consistently.

Why is the time doubled? Since the fetches are async, I would not expect to see any (significant) increase in time.

	import asyncio
	import aiobotocore

	loop = asyncio.get_event_loop()
	session = aiobotocore.get_session(loop=loop)
	s3 = session.create_client('s3')
	    
	async def main():
	    coros = [s3.get_object(Bucket='mybucket', Key=f'{i}.txt') for i in range(5)]
	    await asyncio.gather(*coros)

	def lambda_handler(event, context):
	    loop.run_until_complete(main())

**Comment by thehesiod** [comment]
botocore sets the default connection pool size to 10 (max_pool_connections): https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html#botocore-config, so in aiobotocore we follow the same config option.  Try bumping up the limit.

**Comment by ottokruse** [comment]
Thx for the quick response!

I did see that advice also in another issue but I thought it does not matter in my test as my test compares 5 parallel with 10 parallel gets (so within the pool size). Right?

Anyway I tried it like this also, per your suggestion, but no luck it gives the same results (10 gets take 300ms, 5 gets 150ms):

	import asyncio
	import aiobotocore

	loop = asyncio.get_event_loop()
	session = aiobotocore.get_session(loop=loop)
	config = aiobotocore.config.AioConfig(max_pool_connections=20)
	s3 = session.create_client('s3', config=config)
	    
	async def main():
	    coros = [s3.get_object(Bucket='ottotest', Key=f'{i}.txt') for i in range(10)]
	    await asyncio.gather(*coros)

	def lambda_handler(event, context):
	    loop.run_until_complete(main())

**Comment by thehesiod** [comment]
sorry, ya that's an automatic reply of mine :]  What version of aiobotocore are you using?  Have you checked with a tool like netstat that it's opening 10 connections?  Have you tried locally to see if it's not the lambda environment limiting the number of connections?

**Comment by thehesiod** [comment]
btw, I just tested, and for me it takes ~240ms for either 5 or 10

**Comment by ottokruse** [comment]
Okay, I too did some more testing, this time uploading 1 MB files in parallel. It is pretty hard getting consistent results when testing against the real S3. Therefore I also tried the tests against localstack. Which also gave varying results.

When I logged this issue it seemed that 10 parallel calls were consistently taking twice as long as 5 parallel calls (which would indicate the parallelism/asynchronity wasn't that good) - but I can no longer consistently reproduce that now.

So let's close this issue.

My test result (tried each concurrency 5 times):

	RUNNING AGAINST: realS3
	CONCURRENT_UPLOADS: 20
	AVG: 0.7553291320800781
	STDEV: 0.1336022272480902
	AVG PER FILE: 0.037766456604003906

	CONCURRENT_UPLOADS: 10
	AVG: 0.6403518676757812
	STDEV: 0.07623063751756547
	AVG PER FILE: 0.06403518676757812

	CONCURRENT_UPLOADS: 9
	AVG: 0.693861436843872
	STDEV: 0.21964777226251206
	AVG PER FILE: 0.07709571520487467

	CONCURRENT_UPLOADS: 8
	AVG: 0.5313480854034424
	STDEV: 0.03652483831959031
	AVG PER FILE: 0.0664185106754303

	CONCURRENT_UPLOADS: 7
	AVG: 0.5265211582183837
	STDEV: 0.09271879131704171
	AVG PER FILE: 0.07521730831691197

	CONCURRENT_UPLOADS: 6
	AVG: 0.6723431587219239
	STDEV: 0.26590627082546586
	AVG PER FILE: 0.11205719312032064

	CONCURRENT_UPLOADS: 5
	AVG: 0.6102272033691406
	STDEV: 0.24302141648768005
	AVG PER FILE: 0.12204544067382812

	CONCURRENT_UPLOADS: 4
	AVG: 0.5466031074523926
	STDEV: 0.0583227380867513
	AVG PER FILE: 0.13665077686309815

	CONCURRENT_UPLOADS: 3
	AVG: 0.5162561893463135
	STDEV: 0.15983917607774503
	AVG PER FILE: 0.17208539644877116

	CONCURRENT_UPLOADS: 2
	AVG: 0.4765814781188965
	STDEV: 0.0879081164242467
	AVG PER FILE: 0.23829073905944825

	CONCURRENT_UPLOADS: 1
	AVG: 0.4248361587524414
	STDEV: 0.12345410094298556
	AVG PER FILE: 0.4248361587524414

	----------------------
	RUNNING AGAINST: localstack
	CONCURRENT_UPLOADS: 20
	AVG: 2.1958341121673586
	STDEV: 0.01232173029259046
	AVG PER FILE: 0.10979170560836793

	CONCURRENT_UPLOADS: 10
	AVG: 1.2789833545684814
	STDEV: 0.0013673121032035738
	AVG PER FILE: 0.12789833545684814

	CONCURRENT_UPLOADS: 9
	AVG: 1.4872204303741454
	STDEV: 0.6871222265620456
	AVG PER FILE: 0.16524671448601616

	CONCURRENT_UPLOADS: 8
	AVG: 0.8309378623962402
	STDEV: 0.15216509880999024
	AVG PER FILE: 0.10386723279953003

	CONCURRENT_UPLOADS: 7
	AVG: 0.7101989269256592
	STDEV: 0.1204835320829702
	AVG PER FILE: 0.10145698956080847

	CONCURRENT_UPLOADS: 6
	AVG: 0.7350189685821533
	STDEV: 0.16929777813923458
	AVG PER FILE: 0.12250316143035889

	CONCURRENT_UPLOADS: 5
	AVG: 0.6041313171386719
	STDEV: 0.1748490996507891
	AVG PER FILE: 0.12082626342773437

	CONCURRENT_UPLOADS: 4
	AVG: 0.6443182945251464
	STDEV: 0.34110473238042505
	AVG PER FILE: 0.1610795736312866

	CONCURRENT_UPLOADS: 3
	AVG: 0.7727263450622559
	STDEV: 0.27574160329006464
	AVG PER FILE: 0.2575754483540853

	CONCURRENT_UPLOADS: 2
	AVG: 0.595435094833374
	STDEV: 0.34448011061703243
	AVG PER FILE: 0.297717547416687

	CONCURRENT_UPLOADS: 1
	AVG: 0.495554780960083
	STDEV: 0.2980597547604019
	AVG PER FILE: 0.495554780960083
